### PR TITLE
Add SGX target to rustc

### PR DESCRIPTION
**Pull request by jethrogb**
This adds the `x86_64-fortanix-unknown-sgx` target specification to the Rust compiler. See #56066 for more details about this target.

### CODE REVIEW

File: src/librustc_target/spec/mod.rs [hunk 1]
```diff
@@ -683,6 +685,10 @@ pub struct TargetOptions {
     /// target features. This is `true` by default, and `false` for targets like
     /// wasm32 where the whole program either has simd or not.
     pub simd_types_indirect: bool,
+
+    /// If set, have the linker export exactly these symbols, instead of using
+    /// the usual logic to figure this out from the crate itself.
+    pub override_export_symbols: Option<Vec<String>>
```
Comment (eddyb) [inline src/librustc_target/spec/mod.rs]: cc @alexcrichton

File: src/librustc_codegen_ssa/back/linker.rs [hunk 2]
```diff
@@ -1050,30 +1050,34 @@ impl<'a> Linker for WasmLd<'a> {
...
 
 fn exported_symbols(tcx: TyCtxt, crate_type: CrateType) -> Vec<String> {
-    let mut symbols = Vec::new();
-
-    let export_threshold = symbol_export::crates_export_threshold(&[crate_type]);
-    for &(symbol, level) in tcx.exported_symbols(LOCAL_CRATE).iter() {
-        if level.is_below_threshold(export_threshold) {
-            symbols.push(symbol.symbol_name(tcx).to_string());
+    if let Some(ref exports) = tcx.sess.target.target.options.override_export_symbols {
+        exports.clone()
```
Comment (alexcrichton) [inline src/librustc_codegen_ssa/back/linker.rs]: Could this use a `return` to avoid the indentation of all the following lines?

Comment (jethrogb) [reply src/librustc_codegen_ssa/back/linker.rs]: Certainly. The indentation actually bothered me a lot too.

**Comment by jethrogb** [comment]
Done

**Comment by alexcrichton** [comment]
@bors: r-

reqeusted to work with the libm change!

**Comment by jethrogb** [comment]
Updated to remove libm.a

**Pull request merged**